src/client.js

```diff
@@ -36,17 +36,60 @@ function initSocket() {
 
 global.socket = initSocket();
 
-function createElement(Component, props) {
-  if (Component.fetchData) {
-    Component.fetchData(store.getState, store.dispatch,
-                        props.location, props.params);
+const history = useScroll(createHistory)({routes: createRoutes(routes)});
+syncReduxAndRouter(history, store);
+
+let lastMatchedLocBefore;
+let lastMatchedLocAfter;
+
+history.listenBefore((location, callback) => {
+  const loc = location.pathname + location.search + location.hash;
+  if (lastMatchedLocBefore === loc) {
+    return callback();
   }
-  return React.createElement(Component, props);
-}
+
+  match({routes: routes, location: loc}, (err, redirectLocation, nextState) => {
+    if (!err && nextState) {
+      fetchData(nextState.components, store.getState, store.dispatch,
+        location, nextState.params)
+        .then(() => {
+          lastMatchedLocBefore = loc;
+          callback();
+        })
+        .catch(err2 => {
+          console.error(err2, 'Error while fetching data');
+          callback();
+        });
+    } else {
+      console.log('Location "%s" did not match any routes (listenBefore)', loc);
+      callback();
+    }
+  });
+});
+
+history.listen((location) => {
+  const loc = location.pathname + location.search + location.hash;
+  if (lastMatchedLocAfter === loc) {
+    return;
+  }
+
+  match({routes: routes, location: loc}, (err, redirectLocation, nextState) => {
```

Comment:
Using `createElement` will lead to this: https://github.com/jlongster/react-redux-universal-hot-example/pull/2). Personally, I found it unusable in my project because of the issue above.
Also, I didn't check, but I'm not sure if multiple pushState (even on the same URL) will cause `createElement` to be called multiple times. If that's the case we are nullifying the advantages of your new routing logic. Note: in the code above I explicitly disabled this feature (by comparing old and new location), because I don't think it's needed for this particular use case.